server/routes/serviceProviderReferrals/reportingForm.test.ts

```diff
@@ -0,0 +1,154 @@
+          'to-date-day': '13',
+        })
+
+        const result = await new ReportingForm(request).data()
+
+        expect(result.reportingDateRange!.fromIncludingDate).toEqual('2021-06-12')
+        expect(result.reportingDateRange!.toIncludingDate).toEqual('2021-06-13')
+        expect(result.error).toBeNull()
+      })
+    })
+
+    describe('with invalid data', () => {
+      describe('when the "from" date is after the "to" date', () => {
+        it('returns an error object with a message', async () => {
+          const request = TestUtils.createRequest({
+            'from-date-year': '2021',
+            'from-date-month': '06',
+            'from-date-day': '12',
+            'to-date-year': '2021',
+            'to-date-month': '06',
+            'to-date-day': '11',
+          })
+
+          const result = await new ReportingForm(request).data()
+
+          expect(result.reportingDateRange).toBeNull()
+          expect(result.error!.errors).toContainEqual({
+            errorSummaryLinkedField: 'from-date-day',
+            formFields: ['from-date-day', 'from-date-month', 'from-date-year'],
+            message: 'The "from" date must be before the "to" date',
+          })
+        })
+      })
+
+      describe('when the "from" date is before Day One (01/06/2021)', () => {
+        it('returns an error object with a message', async () => {
+          const request = TestUtils.createRequest({
+            'from-date-year': '2021',
+            'from-date-month': '05',
+            'from-date-day': '12',
+            'to-date-year': '2021',
+            'to-date-month': '06',
+            'to-date-day': '11',
+          })
+
+          const result = await new ReportingForm(request).data()
+
+          expect(result.reportingDateRange).toBeNull()
+          expect(result.error!.errors).toContainEqual({
+            errorSummaryLinkedField: 'from-date-day',
+            formFields: ['from-date-day', 'from-date-month', 'from-date-year'],
+            message: 'Data before 01 June 2021 is not available',
+          })
+        })
+      })
+
+      describe('for dates in the future, for which we have no historical data', () => {
+        const today = new Date()
+        const tomorrow = new Date(today)
+        tomorrow.setDate(tomorrow.getDate() + 1)
+
+        const tomorrowDay = tomorrow.getUTCDate().toString()
+        const tomorrowYear = tomorrow.getUTCFullYear().toString()
+        const tomorrowMonth = (Number(tomorrow.getUTCMonth()) + 1).toString()
+
+        it('returns an error object with a message if the "from" date is in the future', async () => {
+          const request = TestUtils.createRequest({
+            'from-date-year': tomorrowYear,
+            'from-date-month': tomorrowMonth,
+            'from-date-day': tomorrowDay,
+            'to-date-year': '2021',
+            'to-date-month': '06',
+            'to-date-day': '12',
+          })
+
+          const result = await new ReportingForm(request).data()
```

Comment:
It would be easier if you could lock the current time to a known moment here instead of having to do all this formatting logic in the tests. e.g. if you could either pass a `now: Date` instance into the constructor or else stub out `new Date()` (I prefer the former) so that it always returns `new Date('2021-10-21T10:00:00Z')` then you can just check that the error mesage is "Data after 21 October 2021 is not available".

(Although, that said, mightn't it actually be clearer to just say "Data after today is not available"?)

Reply:
I think saying "Data after today is not available" is actually a fair bit clearer (otherwise it makes it sound like the given date is special for some reason). I'm gonna change it to this. It also has the side effect of making the code less complex and resolving two of your comments in one 🥇